File: examples/rest_api/rest_api.py
import requests
import argparse
import json
import os
from aperturedb.CommonLibrary import create_connector
from aperturedb.Connector import Connector

client: Connector = create_connector()

URL = "https://" + client.config.host  + '/api'

VERIFY_SSL = True


def parse_auth(res):

    res = json.loads(res)["json"]
    print(json.dumps(res, indent=4, sort_keys=False))

    session_token = res[0]["Authenticate"]["session_token"]
    refresh_token = res[0]["Authenticate"]["refresh_token"]
    return session_token, refresh_token


def auth():

    query = [{
        "Authenticate": {
            "username": client.config.username,
            "password": client.config.password,
        }
    }]

    # Authenticate
    response = requests.post(URL,
                             files = [('query', (None, json.dumps(query)))],
                             verify = VERIFY_SSL)

    # print(response.status_code)
    # print(response.text)

    return parse_auth(response.text)


def query_api(query, st, files_upload=[]):

    files = [
        ('query', (None, json.dumps(query))),
    ]

    for file in files_upload:
        instream = open(file, 'rb')
        files.append(
            ('blobs', (os.path.basename(file), instream, 'image/jpeg')))

    response = requests.post(URL,
                             headers = {'Authorization': "Bearer " + st},
                             files   = files,
                             verify  = VERIFY_SSL)

    # Parse response:
    try:
        json_response = json.loads(response.text)
        response      = json_response["json"]
        blobs         = json_response["blobs"]
    except:
        print("Error with response:")
        print(response.status_code)
        print(response.text)
        response = "error!"
        blobs = []

    return response, blobs


def get_status(st):

    query = [{
        "GetStatus": {}
    }]

    return query_api(query, st)


def add_image_by_id(st, id):

    query = [{
        "AddImage": {
            "properties": {
                "rest_api_example_id": id
            }
        }
    }]

    return query_api(query, st, files_upload=["songbird.jpg"])


def get_image_by_id(st, id):

    query = [{
        "FindImage": {
            "constraints": {
                "_uniqueid": ["==", id]
            },
            "results": {
                "all_properties": True
            }
        }
    }]

    return query_api(query, st)


def list_images(st):

    query = [{
        "FindImage": {
            "blobs": False,
            "uniqueids": True
        }
    }]

    return query_api(query, st)


def main(params):

    VERIFY_SSL = params.verify_ssl

    print("-" * 80)
    print("Authentication:")
    session_token, refresh_token = auth()

    # Print DB Status
    # get_status(session_token)

    # ----------------------
    print("-" * 80)
    print("List Images:")
    r, blobs = list_images(session_token)
    print("Response:")
    print(json.dumps(r, indent=4, sort_keys=False))
    img_id = r[0]["FindImage"]["entities"][0]["_uniqueid"]

    # ----------------------
    print("-" * 80)
    print("Find image by id:")
    r, blobs = get_image_by_id(session_token, img_id)

    print("Response:")
    print(json.dumps(r, indent=4, sort_keys=False))

    print("Returned images: {}".format(len(blobs)))

    # Base 64 encoded images
    for img in blobs:

        print("Image size (base64 enconded): {}".format(len(img)))

    # ----------------------
    print("-" * 80)
    print("Add image by id:")
    r, blobs = add_image_by_id(session_token, 123456789)

    print("Response:")
    print(json.dumps(r, indent=4, sort_keys=False))

    # ----------------------


def get_args():
    obj = argparse.ArgumentParser()

    obj.add_argument('-verify_ssl',  type=bool, default=True)

    params = obj.parse_args()

    return params


if __name__ == "__main__":
    args = get_args()
    main(args)

File: examples/rest_api/rest_api.js
const request = (query, blobs, handler, sessionToken) => {
    apiURL =  "https://coco.datasets.aperturedata.io/api"
    const formData = new FormData();
    formData.append('query', JSON.stringify(query));
    displayContent(query, response=false);
    
    blobs.forEach(element => {
        formData.append('blobs', element);
    });

    let headers = null;
    if (sessionToken != null){
        console.log(`setting session token ${sessionToken}`);
        headers = {
            "Authorization": `Bearer ${sessionToken}`
        }
    }
    
    axios.post(
            url=apiURL,
            data=formData, {
                headers: headers
            }).then((response)=>{
                handler(response.data)
            })
}

const displayContent = (payload, response=true) => {
    var tag = document.createElement("p");
    var text = JSON.stringify(payload, undefined, 4);
    var element = document.getElementById("output");
    var br = document.createElement("hr");
    prefix = response ? "<<<<<<< Response" : "Request >>>>>>>";
    tag.innerHTML = `${prefix}\r\n${text}`;
    element.appendChild(tag);
    element.appendChild(br);
}

run_requests = () => {
    //Get a refresh token.
    auth = [{
        "Authenticate": {
            "username": "admin",
            "password": "admin"
        }
    }]
    request(query = auth, blobs = [], handler = (data)=>{
        authData = data["json"];
        // console.log(authData[0]);
        displayContent(authData);
        sessionToken = authData[0].Authenticate.session_token;
        

        //List images
        listQuery = [{
            "FindImage": {
                "blobs": false,
                "uniqueids": true,
                "results" : {
                    "limit": 10
                }
            }
        }]
        request(query = listQuery, blobs = [], handler = (data) => {
            response = data["json"];
            displayContent(response);

            //Find an image
            findQuery = [{
                "FindImage": {
                    "constraints": {
                        "_uniqueid": ["==", response[0].FindImage.entities[0]._uniqueid]
                    },
                    "results": {
                        "all_properties": true
                    }
                }
            }]
            request(query = findQuery, blobs = [], handler = (data) => {
                response = data["json"];
                console.log(data);
                displayContent(response);
                const url = `data:image/jpeg;base64,${data["blobs"][0]}`;
                fetch(url)
                .then(res=>res.blob())
                .then(blob=>{
                    var image = document.createElement('img');
                    console.log(blob);
                    image.src = window.webkitURL.createObjectURL(blob);
                    var element = document.getElementById("output");
                    element.appendChild(image);
                });
            }, sessionToken = sessionToken)


        }, sessionToken=sessionToken)

        sessionStorage.setItem("session_token", sessionToken);
    })

    
    
}

const addImage = (event) => {
    event.preventDefault();
    query = [{
        "AddImage": {
            "properties": {
                "rest_api_example_id": 123456789
            }
        }
    }];
    const file = document.getElementById("fileupload").files[0]
    request(query = query, blobs = [file], (data)=>{
        response = data["json"];
        displayContent(response);
    }, sessionToken = sessionStorage.getItem("session_token"));
    
}

window.addEventListener("load", (event)=>{
    console.log("hello world");
    const form = document.getElementById("addimage");
    form.addEventListener('submit', addImage);
})
